feature(loaders): add cql-stress C-S stress tool

muzarski opened the pull request:
Ref: https://github.com/scylladb/cql-stress

cql-stress is a stress tool implemented in Rust. It supports two CLI frontends: scylla-bench and cassandra-stress.

In this commit we introduce a cql-stress-cassandra-stress stress tool to SCT. Currently, the C-S frontend supports 4 basic commands:
- read
- write
- counter_read
- counter_write

## Changes
- docker
  - created docker image definition containing `cql-stress-cassandra-stress` executable
  - pushed docker image to registry
  - updated `defaults/test_default.yaml` accordingly
- loader
  - introduced `CqlStressCassandraStressThread` class deriving from `CassandraStressThread`
    - overridden `create_stress_cmd` and `_run_cs_stress` methods
  - introduced `CqlStressCassandraStressEvent` with critical default severity
  - introduced `CqlStressCassandraStressExporter` which collects the stats on the fly and record them to prometheus
  - added unit tests
- data_dir
  - updated `data_dir/scylla-dash-per-server-nemesis.master.json`. The generated diff looks strange, but the idea was to just copy-paste the existing `cassandra-stress` config for `write` and `read` workloads.

## PR pre-checks (self review)
<!--- PR should be created as Draft, when CI finished and relevant checkboxes selected, add reviewers and then click on "Ready for review" button.-->
<!--- Put an `x` in all the boxes that apply or create PR and then click on all relevant checkboxes: -->
- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [ ] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [x] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [x] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

fruch commented on defaults/test_default.yaml:
```diff
@@ -216,6 +216,7 @@ stress_image:
   kcl: 'scylladb/hydra-loaders:kcl-jdk8-20210526-ShardSyncStrategyType-PERIODIC'
   harry: 'scylladb/hydra-loaders:cassandra-harry-jdk11-20220816'
   latte: 'scylladb/hydra-loaders:latte-rust1_73-20231025'
+  cql-stress-cassandra-stress: 'scylladb/hydra-loaders:cql-stress-cassandra-stress-20231130'
```
It tool itself has some release cadence and versions we can use them and not dates

Also it would be nice if the docker images would be built on the tool repo

Gemini now does it.

fruch commented on sdcm/tester.py:
```diff
@@ -1866,19 +1867,20 @@ def run_stress_cassandra_thread(
-                                          keyspace_name=keyspace_name,
-                                          stop_test_on_failure=stop_test_on_failure,
-                                          params=params or self.params).run()
+        cls = CqlStressCassandraStressThread if "cql-stress" in stress_cmd else CassandraStressThread
```
I would rather have this explicit in its own function

This is a bit confusing to follow

vponomaryov replied:
Agree.
Moreover, `cls` is reserved name for the current class reference.

muzarski commented:
> Do you run this one with a longevity test for a couple of hours, with nemesis ?

No, unfortunately I didn't have an opportunity to do so. I've only tested the solution with two provision test cases:
- https://github.com/scylladb/scylla-cluster-tests#run-test-locally-with-aws-backend
- https://github.com/scylladb/scylla-cluster-tests#run-test-locally-with-docker-backend

I obviously adjusted the `stress_cmd` config parameter to make use of `cql-stress-cassandra-stress`.

I am new to SCT, and I am still not sure about a lot of available config parameters.

Should I create a seperate test-case as in https://github.com/scylladb/scylla-cluster-tests/pull/6735/commits/cec116422c0f3fb373ca2204c8fd31d41f17aa7f? Should I create a jenkins pipeline for that? Which nemesis should be used? What SCT backend should I use?

I'd appreciate any tips regarding preparation (and running) the test case that would allow us to compare both (cql-stress vs c-s) of the solutions.

fruch commented:
Since it's a drop-in, I would suggest taking a current case that can support it, and duplicate it to use this tool

And then run both on Jenkins under scylla-staging, so results are visible to others

@soyacz can you help him with showing how to reuse existing tests with override configuration ?

soyacz commented:
> 
> @soyacz can you help him with showing how to reuse existing tests with override configuration ?

Sure, @muzarski ping me on the Slack we can have call so I can answer also some other questions

vponomaryov commented on sdcm/loader.py:
```diff
@@ -202,6 +202,57 @@ def split_line(self, line: str) -> list:
+
+        return self.do_skip
+
+    @staticmethod
+    def split_line(line: str) -> list:
+        return [element.strip() for element in line.split(',')]
```
At first, it is unused.
At second, it is not specific to this class.
So, if you want such utility function/method then need to place in some of the `utils` modules.

muzarski replied:
It's used by the parent class `StressExporter`. The default implementation is empty.

fruch replied:
this is the StressExporter api, that's o.k. 

if the output of this tool is identical to c-s, it might be o.k. to inherit from `CassandraStressExporter`, and just replace `create_metrix_gauge`

anyhow it's a nitpick from my POV.

vponomaryov commented on sdcm/cql_stress_cassandra_stress_thread.py:
```diff
@@ -0,0 +1,158 @@
+LOGGER = logging.getLogger(__name__)
+
+
+class CqlStressCassandraStressThread(CassandraStressThread):
```
Why not `CqlStressThread` ?

fruch replied:
also big portion of `CassandraStressThread` implementation is copied here, 
so maybe there no reason to inherit from it at all ?

if we do inherit, we should move the relevant part which needs to be different to their own functions/properties that can be overridden

muzarski replied:
> Why not `CqlStressThread` ?

As `cql-stress` supports two frontends (`cassandra-stress` and `scylla-bench`), I preferred to be more specific.

vponomaryov commented on sdcm/cql_stress_cassandra_stress_thread.py:
```diff
@@ -0,0 +1,158 @@
+
+        if self.node_list and '-node' not in stress_cmd:
+            stress_cmd += " -node "
+            if self.loader_set.test_config.MULTI_REGION:
```
It should consider the SCT config option called `region_aware_loader`.

fruch replied:
this looks like  a code that should be shared, is looks this was copied from cassandra-stress implementation.
and shouldn't (also there an open PR to extend this logic for rack aware configuration)

vponomaryov commented on sdcm/cql_stress_cassandra_stress_thread.py:
```diff
@@ -0,0 +1,158 @@
+                    LOGGER.error("Not found datacenter for loader region '%s'. Datacenter per loader dict: %s",
+                                 loader.region, datacenter_name_per_region)
+
+            node_ip_list = [n.cql_address for n in self.node_list]
+            stress_cmd += ",".join(node_ip_list)
```
It could be simplified to the following:

```
stress_cmd += ",".join((n.cql_address for n in self.node_list))
```

vponomaryov commented on sdcm/cql_stress_cassandra_stress_thread.py:
```diff
@@ -0,0 +1,158 @@
+                                r' throttle=\1 fixed', stress_cmd)
+
+        if '-pop' in stress_cmd:
+            # '-pop seq=x..y' syntax is not YET supported by cql-stress.
+            # TODO remove when seq=x..y syntax is supported.
+            stress_cmd = re.sub(r' seq=[\s]*([\d]+\.\.[\d]+)',
+                                r" 'dist=SEQ(\1)'", stress_cmd)
```
This part worth adding a `warning` log message.

vponomaryov commented on sdcm/cql_stress_cassandra_stress_thread.py:
```diff
@@ -0,0 +1,158 @@
+        stress_cmd = self.stress_cmd
+
+        if "no-warmup" not in stress_cmd:
+            # add no-warmup to stress_cmd if it's not there. See issue #5767
```
Which project's issue?

fruch replied:
this comment seem to be copied from somewhere else...
I don't know what the issue, but we shouldn't be using warmup, it's doesn't have any meaning for scylla, its something todo with cassandra java implementation ontop of the JVM, that need time to "stabilize"

vponomaryov commented on sdcm/cql_stress_cassandra_stress_thread.py:
```diff
@@ -0,0 +1,158 @@
+            "cql-stress-cassandra-stress", 1)[1].split(None, 1)[0]
+
+        log_id = self._build_log_file_id(loader_idx, cpu_idx, keyspace_idx)
+        log_file_name = \
+            os.path.join(
+                loader.logdir, f'cql-stress-cassandra-stress-{stress_cmd_opt}-{log_id}.log')
```
Why not like following:

```
log_file_name = os.path.join(
    loader.logdir, f'cql-stress-cassandra-stress-{stress_cmd_opt}-{log_id}.log')
```

?

vponomaryov commented:
@muzarski I see that you uploaded new changes but haven't posted any inline comment.
So, either you haven't written any or haven't published the `draft` comments. Which one is true?

muzarski commented:
> @muzarski I see that you uploaded new changes but haven't posted any inline comment. So, either you haven't written any or haven't published the `draft` comments. Which one is true?

I haven't pushed any commits regarding the review comments yet. I didn't have time to go through the comments yet.

The commits I've pushed are just draft commits. I'll re-request the review once the code is ready for it.

muzarski commented:
Sorry for the delay, I had to adjust some things to make the longevity test run correctly. Here is the jenkins pipeline: https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/muzarki/job/longevity-100gb-4h-test-cql-stress/.

Note that the last run failed, but as I spoke with @soyacz, the issue is not related with cql-stress.

I'll start fixing the review comments now.

muzarski commented:
v2:
- rebased on top of master
- fixed review comments, including:
  - created separate `run_cql_stress_cassamdra_thread` function
  - extracted common logic of `CassandraStressThread::create_stress_cmd` to functions so they can be reused by `CqlStressCassandraStressThread`

muzarski commented:
v3:
- initiate `result` variable before `with` block in `_run_cs_stress`

fruch commented:
@muzarski 

I'm looking at the dashboard and results from:
https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/muzarki/job/longevity-100gb-4h-test-cql-stress/11

![image](https://github.com/scylladb/scylla-cluster-tests/assets/340979/06d69a58-8e11-413b-93f5-749979a7d160)

1) this dashboard looks duplicated
2) seems like metrics dropped at ~13:00
3) seem to corrlate to logs in 
```
[2m2024-01-05T13:09:31.237763Z[0m [31mERROR[0m [2mcql_stress_cassandra_stress::operation::write[0m[2m:[0m write error [3merror[0m[2m=[0mInvalid message: Frame error: early eof [3mpartition_key[0m[2m=[0mBlob([76, 79, 55, 77, 53, 57, 56, 76, 77, 48])
```

can we drop the ansi colors ? 

I'm guess `early eof` is cause of a node going down, we'll need to be able to tell a bit better which line is critical to stop the test (for example validation error, or failure after x retries)
see `CassandraStressEventsPublisher` for how it's used there

need to check why those lines are breaking the reporting code

without those stats working, it's a bit hard to compare it to regular c-s results/runs

muzarski commented:
> @muzarski
> 
>     1. this dashboard looks duplicated
I'll deduplicate it.

>     2. seems like metrics dropped at ~13:00
> 
>     3. seem to corrlate to logs in
> 
> 
> ```
> �[2m2024-01-05T13:09:31.237763Z�[0m �[31mERROR�[0m �[2mcql_stress_cassandra_stress::operation::write�[0m�[2m:�[0m write error �[3merror�[0m�[2m=�[0mInvalid message: Frame error: early eof �[3mpartition_key�[0m�[2m=�[0mBlob([76, 79, 55, 77, 53, 57, 56, 76, 77, 48])
> ```

It might be because `StressExporter` api failed to parse the `early eof` line. I'll adjust the `skip_line` logic so we skip the lines emitted by [tracing](https://docs.rs/tracing/latest/tracing/) crate.

> can we drop the ansi colors ?

Opened a PR: https://github.com/scylladb/cql-stress/pull/66.

> I'm guess `early eof` is cause of a node going down, we'll need to be able to tell a bit better which line is critical to stop the test (for example validation error, or failure after x retries) see `CassandraStressEventsPublisher` for how it's used there

Yes, I believe the `early eof` error is caused by `disrupt_stop_wait_start_scylla_server` nemesis. Some operations need to be retried to be successful in this case.

I believe that the only `CRITICAL`  event which should stop the test is the read validation error. Are there any other critical events we should consider?

fruch commented:
> > @muzarski
> > 
> >     1. this dashboard looks duplicated
> I'll deduplicate it.
> 
> >     2. seems like metrics dropped at ~13:00
> > 
> >     3. seem to corrlate to logs in
> > 
> > 
> > ```
> > �[2m2024-01-05T13:09:31.237763Z�[0m �[31mERROR�[0m �[2mcql_stress_cassandra_stress::operation::write�[0m�[2m:�[0m write error �[3merror�[0m�[2m=�[0mInvalid message: Frame error: early eof �[3mpartition_key�[0m�[2m=�[0mBlob([76, 79, 55, 77, 53, 57, 56, 76, 77, 48])
> > ```
> 
> It might be because `StressExporter` api failed to parse the `early eof` line. I'll adjust the `skip_line` logic so we skip the lines emitted by [tracing](https://docs.rs/tracing/latest/tracing/) crate.
> 
> > can we drop the ansi colors ?
> 
> Opened a PR: https://github.com/scylladb/cql-stress/pull/66.
> 
> > I'm guess `early eof` is cause of a node going down, we'll need to be able to tell a bit better which line is critical to stop the test (for example validation error, or failure after x retries) see `CassandraStressEventsPublisher` for how it's used there
> 
> Yes, I believe the `early eof` error is caused by `disrupt_stop_wait_start_scylla_server` nemesis. Some operations need to be retried to be successful in this case.
> 
> I believe that the only `CRITICAL`  event which should stop the test is the read validation error. Are there any other critical events we should consider?

I don't have nothing in mind right now, but once we'll have more mileage with this tool, we will probably have some more things we'll want to raise out of the log.

muzarski commented:
v3:
- deduplicated `cql-stress-cassandra-stress ops` dashboard
- added `CqlStressCassandraStressEventsPublisher` which checks for `read validation error` events. See exemplary log from `cql-stress` when row validation fails: `2024-01-18T15:57:52.108866Z ERROR cql_stress_cassandra_stress::operation::read: read validation error error=Query result doesn't contain any rows. partition_key=Blob([75, 75, 79, 51, 76, 48, 57, 48, 49, 49])`
- Adjusted `CqlStressCassandraStressExporter` so it ignores lines emitted by `tracing` crate. It means - to ignore lines containing any of the `TRACE`, `DEBUG`, `INFO`, `WARN`, `ERROR` keywords.

muzarski commented:
Scheduled a SCT run with new changes: https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/muzarki/job/longevity-100gb-4h-test-cql-stress/12/.

roydahan commented:
Any progress with this one?

muzarski commented:
@fruch I addressed review comments and scheduled a run (which was a success)

muzarski commented:
> key observation:
> 
>     1. cql-stress is doing more then x2 requests !
> 
>     2. latency of p99 is below ~2ms vs. ~10ms in c-s

In addition, notice that `cql-stress` workload uses `rate threads=30`, when c-s workload uses `rate threads=40`. I didn't want to overload the cluster so decided to use a smaller number of threads here (actually, this is not the number of threads, but number of async tokio tasks in the pool).

> seems like we still have some duplication in the p99 and p95 graphs

Looking at the photos, I see that we have dashboards for both c-s and cql-stress (C-S frontend). Note that some nemesis could run cassandra-stress in the background and so c-s metrics were collected as well.

fruch commented:
> > key observation:
> > 
> >     1. cql-stress is doing more then x2 requests !
> > 
> >     2. latency of p99 is below ~2ms vs. ~10ms in c-s
> 
> In addition, notice that `cql-stress` workload uses `rate threads=30`, when c-s workload uses `rate threads=40`. I didn't want to overload the cluster so decided to use a smaller number of threads here (actually, this is not the number of threads, but number of async tokio tasks in the pool).

That's something we'll need todo per use case and it's requirements, for this one hitting 50% CPU look o.k.
> 
> > seems like we still have some duplication in the p99 and p95 graphs
> 
> Looking at the photos, I see that we have dashboards for both c-s and cql-stress (C-S frontend). Note that some nemesis could run cassandra-stress in the background and so c-s metrics were collected as well.

Yeah, I've noticed it, but still we have cql-stress read panels duplicated here

Anyhow that's good enough to start using in some of other cases

> 
>

muzarski commented:
@fruch resolved the conflict

fruch reviewed (approved):
LGTM

roydahan commented:
Great work @muzarski!
I was happy to see that you took care also the grafana dashboards and was able to run and demonstrate everything.

roydahan commented:
@fruch I see that the new pipeline was added under "rust" folder.
This is probably not handled in our tool that creates the jobs in Jenkins automatically.
Is your PR for "pipelines" re-org going to take care of that or we need to manually fix it now?
